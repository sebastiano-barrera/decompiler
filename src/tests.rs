#![cfg(test)]

mod callconv_x86_64 {
    use crate::{mil, ssa, ty, x86_to_mil};

    use iced_x86::code_asm::CodeAssembler;
    use iced_x86::Instruction;

    #[test]
    fn struct_2p() {
        let input = {
            use iced_x86::code_asm::{eax, ptr, rsi};
            let mut asm = CodeAssembler::new(64).unwrap();

            let mut lbl_end = asm.create_label();
            let mut lbl_loop = asm.create_label();

            asm.mov(eax, 1).unwrap();
            asm.test(rsi, rsi).unwrap();
            asm.je(lbl_end).unwrap();
            asm.set_label(&mut lbl_loop).unwrap();
            asm.inc(eax).unwrap();
            asm.mov(rsi, ptr(rsi + 8)).unwrap();
            asm.test(rsi, rsi).unwrap();
            asm.jne(lbl_loop).unwrap();
            asm.set_label(&mut lbl_end).unwrap();
            asm.ret().unwrap();

            asm.take_instructions()
        };

        let mut types = ty::TypeSet::new();
        let tyid_char = types.add(ty::Type {
            name: "char".to_owned().into(),
            ty: ty::Ty::Int(ty::Int {
                size: 1,
                signed: ty::Signedness::Signed,
            }),
        });
        let tyid_char_ptr = types.add(ty::Type {
            name: "char*".to_owned().into(),
            ty: ty::Ty::Ptr(tyid_char),
        });
        let tyid_name_item = {
            let tyid = types.add(ty::Type {
                name: "name_item".to_owned().into(),
                ty: ty::Ty::Void,
            });
            let self_ptr = types.add(ty::Type {
                name: "name_item*".to_owned().into(),
                ty: ty::Ty::Ptr(tyid),
            });

            types.set(
                tyid,
                ty::Type {
                    name: "name_item".to_owned().into(),
                    ty: ty::Ty::Struct(ty::Struct {
                        size: 16,
                        members: vec![
                            ty::StructMember {
                                offset: 0,
                                name: "name".to_owned().into(),
                                tyid: tyid_char_ptr,
                            },
                            ty::StructMember {
                                offset: 0,
                                name: "next".to_owned().into(),
                                tyid: self_ptr,
                            },
                        ],
                    }),
                },
            );

            tyid
        };

        print_asm(&input);
        let prog = {
            let mut b = x86_to_mil::Builder::new();
            b.use_types(&types);
            b.read_func_args(&[tyid_name_item]).unwrap();
            b.translate(input.iter().copied())
        }
        .unwrap();
        let output = finish_prog(prog).unwrap();
        insta::assert_snapshot!(output);
    }

    fn finish_prog(prog: mil::Program) -> Result<String, std::fmt::Error> {
        use std::fmt::Write;

        let mut out = String::new();
        writeln!(out)?;
        writeln!(out, "mil program = ")?;
        writeln!(out, "{:?}", prog)?;

        let mut prog = ssa::mil_to_ssa(ssa::ConversionParams::new(prog));
        crate::xform::canonical(&mut prog);
        ssa::eliminate_dead_code(&mut prog);
        writeln!(out)?;
        writeln!(out, "{:?}", prog)?;

        // TODO print AST as well?

        // let out = std::io::stdout().lock();
        // let mut pp = PrettyPrinter::start(IoAsFmt(out));

        // println!();
        // let ast = ast::ssa_to_ast(&prog);
        // ast.pretty_print(&mut pp).unwrap()

        Ok(out)
    }

    fn print_asm(instrs: &[Instruction]) {
        use iced_x86::{Formatter, IntelFormatter};
        let mut formatter = IntelFormatter::new();
        let mut instr_strbuf = String::new();
        for instr in instrs {
            print!("{:16x}: ", instr.ip());
            instr_strbuf.clear();
            formatter.format(&instr, &mut instr_strbuf);
            println!("{}", instr_strbuf);
        }
    }
}
