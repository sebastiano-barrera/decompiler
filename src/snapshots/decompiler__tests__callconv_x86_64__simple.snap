---
source: src/tests.rs
expression: output
---
mil program = 
program  72 instrs
    0 r12 <- #pre:stack_bottom
    1 r2 <- #pre:CF
    2 r3 <- #pre:PF
    3 r4 <- #pre:AF
    4 r5 <- #pre:ZF
    5 r6 <- #pre:SF
    6 r7 <- #pre:TF
    7 r8 <- #pre:IF
    8 r9 <- #pre:DF
    9 r10 <- #pre:OF
   10 r11 <- #pre:RBP
   11 r12 <- #pre:RSP
   12 r13 <- #pre:RIP
   13 r14 <- #pre:RDI
   14 r15 <- #pre:RSI
   15 r16 <- #pre:RAX
   16 r17 <- #pre:RBX
   17 r18 <- #pre:RCX
   18 r19 <- #pre:RDX
   19 r20 <- #pre:R8
   20 r21 <- #pre:R9
   21 r22 <- #pre:R10
   22 r23 <- #pre:R11
   23 r24 <- #pre:R12
   24 r25 <- #pre:R13
   25 r26 <- #pre:R14
   26 r27 <- #pre:R15
   27 r28 <- #pre:arg0
   28 r29 <- sget8    r28,0
   29 r14 <- get      r29
   30 r30 <- sget8    r28,8
   31 r15 <- get      r30
   32 r28 <- undef
   33 r29 <- undef
   34 r30 <- undef
   35 r31 <- undef
   36 r32 <- undef
   37 r33 <- undef
   38 r34 <- undef
   39 r28 <- const4   1 (0x1)
   40 r16 <- v8.l4=   r16 ← r28
   41 r15 <- and      r15,r15  qword
   42 r6 <- sign     r15
   43 r5 <- is0      r15
   44 r30 <- l1       r15
   45 r3 <- parity   r15
   46 r2 <- const1   0 (0x0)
   47 r10 <- const1   0 (0x0)
   48 r28 <- jmp.if   r5,71
0x2:
   49 r28 <- l4       r16
   50 r28 <- addk     r28,1 dword
   51 r16 <- v8.l4=   r16 ← r28
   52 r10 <- false
   53 r2 <- false
   54 r6 <- sign     r28
   55 r5 <- is0      r28
   56 r29 <- l1       r28
   57 r3 <- parity   r29
0x0:
   58 r29 <- const8   8 (0x8)
   59 r29 <- add      r29,r15  qword
   60 r28 <- loadm8   addr:r29
   61 r15 <- get      r28
   62 r15 <- and      r15,r15  qword
   63 r6 <- sign     r15
   64 r5 <- is0      r15
   65 r30 <- l1       r15
   66 r3 <- parity   r15
   67 r2 <- const1   0 (0x0)
   68 r10 <- const1   0 (0x0)
   69 r28 <- not      r5
   70 r29 <- jmp.if   r28,49
0x1:
   71 r28 <- ret      r16


ssa program  20 instrs
.B0:    ;;  addr:0x0; 49 insn 0 phis
  (  2)  r15 <- #pre:RAX
         r27 <- #pre:arg0
  (  2)  r30 <- sget8    r27,8
  (  2)  r39 <- const4   1 (0x1)
  (  3)  r40 <- v8.l4=   r15 ← r39
         r41 <- get      r30
         r43 <- is0      r30
         r48 <- jmp.if   r43,71
.B1:    ;;  preds:B0,B1  addr:0x2; 22 insn 2 phis
                  ɸ  B0     B1     
           r72    <- r41    r62    
  (  2)    r75    <- r40    r51    
         r50 <- addk     r75,1 dword
  (  3)  r51 <- v8.l4=   r75 ← r50
         r59 <- addk     r72,8 qword
  (  2)  r60 <- loadm8   addr:r59
         r62 <- get      r60
         r64 <- is0      r60
         r69 <- not      r64
         r70 <- jmp.if   r69,49
.B2:    ;;  preds:B0,B1  addr:0x1; 1 insn 1 phis
                  ɸ  B0     B1     
           r78    <- r40    r51    
         r71 <- ret      r78
